src/py42/clients/settings/device_settings.py

```diff
@@ -189,38 +195,51 @@ def _extract_file_selection_lists(self):
         """
-        pathset = self.data[u"backupPaths"][u"pathset"][0]
-        pathlist = pathset.get(u"path")
+        pathset = self.data[u"backupPaths"][u"pathset"]
+        if isinstance(pathset, dict):  # pathset is locked
+            path_list = pathset[u"paths"].get(u"path")
+        else:
+            path_list = pathset[0].get(u"path")
 
         # no paths selected
-        if pathlist is None:
+        if path_list is None:
             return [], []
 
         # one path selected
-        if isinstance(pathlist, dict):
-            pathlist = [pathlist]
+        if isinstance(path_list, dict):
+            path_list = [path_list]
 
-        includes = [p[u"@include"] for p in pathlist if u"@include" in p]
-        excludes = [p[u"@exclude"] for p in pathlist if u"@exclude" in p]
+        includes = [p[u"@include"] for p in path_list if u"@include" in p]
+        excludes = [p[u"@exclude"] for p in path_list if u"@exclude" in p]
         return includes, excludes
 
     def _extract_regex_exclusions(self):
         """Converts the filename exclusion portion ("excludeUser") of the settings dict
         into a simple list of regex patterns.
 
         The "excludeUser" object is a different shape based on the number of exclusion
-        patterns present:
-            No exclusions:   `[{"windows": [], "linux": [], "macintosh": []}]`
-            One exclusion:   `[{"windows": [], "pattern": {"@regex": ".*"}, "linux": [], "macintosh": []}]`
-            One+ exclusions: `[{"windows": [], "pattern": [{"@regex": ".*1"}, {"@regex": ".*2"}],"linux": [],"macintosh": []}]
+        patterns present and whether the setting is locked or not:
+            No exclusions:          `[{"windows": [], "linux": [], "macintosh": []}]`
+            No exclusions locked:   `{'@locked': 'true', 'patternList': {'windows': [], 'macintosh': [], 'linux': []}}`
+            One exclusion:          `[{"windows": [], "pattern": {"@regex": ".*"}, "linux": [], "macintosh": []}]`
+            One exclusion locked:   `{'@locked': 'true', 'patternList': {'pattern': {'@regex': '.*'}, 'windows': [], 'macintosh': [], 'linux': []}}`
+            One+ exclusions:        `[{"windows": [], "pattern": [{"@regex": ".*1"}, {"@regex": ".*2"}],"linux": [],"macintosh": []}]
+            One+ exclusion locked:  `{'@locked': 'true', 'patternList': {'pattern': [{'@regex': '.*1'}, {'@regex': '.*2'}], 'windows': [], 'macintosh': [], 'linux': []}}`
         """
-        exclude_user = self.data[u"backupPaths"][u"excludeUser"][0]
-        pattern_list = exclude_user.get(u"pattern")
+        exclude_user = self.data[u"backupPaths"][u"excludeUser"]
+        if isinstance(exclude_user, dict):  # exclusions are locked
+            pattern_list = exclude_user[u"patternList"].get(u"pattern")
+        else:
+            pattern_list = exclude_user[0].get(u"pattern")
```

Comment:
`No exclusions` sample above doesn't contain "pattern" key, so `pattern_list` would be empty for `No exclusions`, is it expected?

Reply:
I was assuming the locks are in place because they are set at the org level, or is that not it?

Reply:
> it would have been better that the API would implement such logic and we didn't have to bother for such stuff.

yes.

> Do we really want to fail early? or can we push each task to a queue and kind of guarantee that each update would be successful, where every task waits for the lock to be released by earlier task for certain number of attempts and only then fail if it doesn't acquire the lock, and so on for all tasks in the queue.

@timabrmsn  should probably confirm but my understanding is that this isn't a _concurrency_ lock (that is, its not a lock that only exists while someone else is making modifications but then later goes away). I think this is a _restrictive_ lock, i.e. an admin disabling the ability of other users to modify this setting.

Reply:
> pattern_list would be empty for No exclusions, is it expected?
yep, if there aren't any exclusion patterns we just want to return an empty list. 

And for locking, @alanag13 and @unparalleled-js are correct, it's an administrative restriction on the device settings, not a temporary lock because changes are in flight. 

And what makes it even more complex is that each individual item can be locked/unlocked, and the _whole backup set_ can be locked, which has the effect of overriding any individual lock settings. It's only in the case of the whole backup set being locked that we fail fast and just not let the attribute be changed, since it's possible to update those values in the json and post it in in the API, but if the whole set is locked the server will just ignore any changes you make, so we're calling attention to the fact that they'd _have_ to unlock the backup set first before making any changes.